wvd-templates/DSC/Functions.ps1

```diff
@@ -222,13 +222,159 @@ function ExtractDeploymentAgentZipFile {
+        $tenants = Get-RdsTenant -Name "$TenantName"
+        if (!$tenants) {
+            Write-Log "No tenants exist or you do not have proper access."
+            #//todo throw ?
+        }
+    }
+    catch {
+        #//todo refactor msg ?
+        Write-Log -Message $_
+        throw $_
+    }
+}
+
+function ImportRDPSMod {
+    param(
+        [string]$Source = 'attached',
+        [string]$ArtifactsPath
+
+        # [Parameter(mandatory = $false)]
+        # [string]$DeployAgentLocation = 'C:\DeployAgent'
+    )
+
+    #//todo move outside
+    # Write-Log -Message "Creating a folder inside rdsh vm for extracting deployagent zip file"
+    # ExtractDeploymentAgentZipFile -ScriptPath $ScriptPath -DeployAgentLocation $DeployAgentLocation
+    
+    # Write-Log -Message "Changing current folder to Deployagent folder: $DeployAgentLocation"
+    # Set-Location "$DeployAgentLocation"
+    
+    if ($Source -eq 'attached') {
+        if ((-not $ArtifactsPath) -or (-not (test-path $ArtifactsPath))) {
+            throw "invalid param: ArtifactsPath = '$ArtifactsPath'"
+        }
+
+        # Locating and extracting PowerShellModules.zip
+        $ZipName = 'PowerShellModules.zip'
+        Write-Log -Message "Locating '$ZipName' within Custom Script Extension folder structure: '$ArtifactsPath'"
+        $ZipPath = (Get-ChildItem "$ArtifactsPath\" -Filter $ZipName -Recurse | Select-Object -First 1).FullName
+        if ((-not $ZipPath) -or (-not (Test-Path $ZipPath))) {
+            throw "'$ZipName' file not found at '$ArtifactsPath'"
+        }
+        
+        $Path = 'C:\PowerShellModules\'
+        if (test-path $Path) {
+            Remove-Item -Path $Path -Force -Recurse -ErrorAction 'SilentlyContinue'
+        }
+
+        Write-Log -Message "Extracting '$ZipPath' file into '$Path' folder inside VM"
+        Expand-Archive $ZipPath -DestinationPath $Path -Force -ErrorAction 'SilentlyContinue'
+
+        $DLLName = 'Microsoft.RDInfra.RDPowershell.dll'
+        Write-Log -Message "Locating '$DLLName' within '$Path'"
+        $DLLPath = (Get-ChildItem "$Path\" -Filter $DLLName -Recurse | Select-Object -First 1).FullName
+        if ((-not $DLLPath) -or (-not (Test-Path $DLLPath))) {
+            throw "'$DLLName' file not found at '$Path'"
+        }
+
+        Write-Log -Message "Importing WVD PowerShell modules from attached artifacts"
+        Import-Module $DLLPath
+        Write-Log -Message "Imported Windows Virtual Desktop PowerShell modules successfully from attached artifacts"
+        return
+    }
+
+    $Version = ($Source.Trim().ToLower() -split 'gallery@')[1]
+    if ($Version -eq $null -or $Version.Trim() -eq '') {
+        throw "invalid param: Source = $Source"
+    }
+
+    Write-Log -Message "Installing & importing WVD PowerShell modules (version: v$Version) from PowerShell Gallery"
+    if ($Version -eq 'latest') {
+        Install-Module -Name Microsoft.RDInfra.RDPowershell -Scope 'CurrentUser' -AllowClobber -Force

```

Comment:
I think we may want to consider using Save-Module (which saves the powershell module to a local folder) and then importing it just like we do for the bundled module version instead of Install-Module. I assume that Install-Module will leave the installed module behind for the current user after this DSC finished but using Save-Module won't do that.
Here is an example of using Save-Module
`Save-Module -Name Microsoft.RDInfra.RDPowershell -Path "E:\git\test" -RequiredVersion (new-object System.Version("1.0.833")) -Force `

Reply:
I think it's fine to leave the module installed because the user will know the implication of it as the user has to manually modify the DSC to force use the gallery anyways. What do you think @ChristianMontoya ?

Reply:
Do we know if having the RDPowershell  module installed could conflict with the RDAgent if that installs a different version of RDPowershell ?

Reply:
Oh I didn't know that RDAgent also installs the RDPowerShell module.

Reply:
I don't know if it does but it could. Another thing to consider is that the purpose of this is to let the user run this template with a specific version of the powershell module rather than install it on the machine, this current code doesn't let the user choose if the module should be installed for all users or the current user and we also don't have a good way to show installation errors to the user, such as if there is already a different version of it installed normally they would get a warning about that if they didn't specify -AllowClobber -Force on the Install-Module command. So it doesn't seem like it is a very useful feature to include leaving it installed.